### PR TITLE
simplify the `between` expr during logical plan optimization

### LINKED ISSUE #3402: add simplification for `between` expression during logical plan optimization
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Take TPC-H q6 for example.
Plan prior to simplification:
* has a strange "AND true" clause in the ParquetExec
* not able to push l_discount down into the ParquetExec
```
❯ explain select
    sum(l_extendedprice * l_discount) as revenue
from
    lineitem
where
        l_shipdate >= date '1994-01-01'
  and l_shipdate < date '1995-01-01'
  and l_discount between 0.06 - 0.01 and 0.06 + 0.01
  and l_quantity < 24;
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan
                                                                                                                                              |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #SUM(lineitem.l_extendedprice * lineitem.l_discount) AS revenue                                                                                                                                              |
|               |   Aggregate: groupBy=[[]], aggr=[[SUM(#lineitem.l_extendedprice * #lineitem.l_discount)]]                                                                                                                                              |
|               |     Filter: #lineitem.l_shipdate >= Date32("8766") AND #lineitem.l_shipdate < Date32("9131") AND #lineitem.l_discount BETWEEN Float64(0.049999999999999996) AND Float64(0.06999999999999999) AND #lineitem.l_quantity < Decimal128(Some(2400),15,2)                                                                                          |
|               |       TableScan: lineitem projection=[l_quantity, l_extendedprice, l_discount, l_shipdate], partial_filters=[#lineitem.l_shipdate >= Date32("8766"), #lineitem.l_shipdate < Date32("9131"), #lineitem.l_discount BETWEEN Float64(0.049999999999999996) AND Float64(0.06999999999999999), #lineitem.l_quantity < Decimal128(Some(2400),15,2)] |
| physical_plan | ProjectionExec: expr=[SUM(lineitem.l_extendedprice * lineitem.l_discount)@0 as revenue]                                                                                                                                              |
|               |   AggregateExec: mode=Final, gby=[], aggr=[SUM(lineitem.l_extendedprice * lineitem.l_discount)]                                                                                                                                              |
|               |     CoalescePartitionsExec                                                                                                                                              |
|               |       AggregateExec: mode=Partial, gby=[], aggr=[SUM(lineitem.l_extendedprice * lineitem.l_discount)]                                                                                                                                              |
|               |         CoalesceBatchesExec: target_batch_size=4096                                                                                                                                              |
|               |           FilterExec: l_shipdate@3 >= 8766 AND l_shipdate@3 < 9131 AND CAST(l_discount@2 AS Decimal128(30, 15)) >= CAST(0.049999999999999996 AS Decimal128(30, 15)) AND CAST(l_discount@2 AS Decimal128(30, 15)) <= CAST(0.06999999999999999 AS Decimal128(30, 15)) AND l_quantity@0 < Some(2400),15,2                                       |
|               |             RepartitionExec: partitioning=RoundRobinBatch(20)                                                                                                                                              |
|               |               ParquetExec: limit=None, partitions=[home/kmitchener/dev/arrow-datafusion/benchmarks/data-parquet/lineitem/part-0.parquet], predicate=l_shipdate_max@0 >= 8766 AND l_shipdate_min@1 < 9131 AND true AND l_quantity_min@2 < Some(2400),15,2, projection=[l_quantity, l_extendedprice, l_discount, l_shipdate]                   |
|               |                                                                                                                                              |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```
**Describe the solution you'd like**

Convert `between` expression into >= and <= expressions and there is more opportunity for further optimize it in the logical plan.
It results in a better plan overall, with more predicates pushed down to the tablescan.

```
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                           |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #SUM(lineitem.l_extendedprice * lineitem.l_discount) AS revenue                                                                                                           |
|               |   Aggregate: groupBy=[[]], aggr=[[SUM(#lineitem.l_extendedprice * #lineitem.l_discount)]]                                                                                                           |
|               |     Filter: #lineitem.l_shipdate >= Date32("8766") AND #lineitem.l_shipdate < Date32("9131") AND CAST(#lineitem.l_discount AS Decimal128(30, 15)) >= CAST(Float64(0.049999999999999996) AS Decimal128(30, 15)) AND CAST(#lineitem.l_discount AS Decimal128(30, 15)) <= CAST(Float64(0.06999999999999999) AS Decimal128(30, 15)) AND #lineitem.l_quantity < Decimal128(Some(2400),15,2)                                                                                                           |
|               |       TableScan: lineitem projection=[l_quantity, l_extendedprice, l_discount, l_shipdate], partial_filters=[#lineitem.l_shipdate >= Date32("8766"), #lineitem.l_shipdate < Date32("9131"), #lineitem.l_discount >= Float64(0.049999999999999996), #lineitem.l_discount <= Float64(0.06999999999999999), #lineitem.l_quantity < Decimal128(Some(2400),15,2)]                                                                                                           |
| physical_plan | ProjectionExec: expr=[SUM(lineitem.l_extendedprice * lineitem.l_discount)@0 as revenue]                                                                                                           |
|               |   AggregateExec: mode=Final, gby=[], aggr=[SUM(lineitem.l_extendedprice * lineitem.l_discount)]                                                                                                           |
|               |     CoalescePartitionsExec                                                                                                           |
|               |       AggregateExec: mode=Partial, gby=[], aggr=[SUM(lineitem.l_extendedprice * lineitem.l_discount)]                                                                                                           |
|               |         CoalesceBatchesExec: target_batch_size=4096                                                                                                           |
|               |           FilterExec: l_shipdate@3 >= 8766 AND l_shipdate@3 < 9131 AND CAST(l_discount@2 AS Decimal128(30, 15)) >= CAST(0.049999999999999996 AS Decimal128(30, 15)) AND CAST(l_discount@2 AS Decimal128(30, 15)) <= CAST(0.06999999999999999 AS Decimal128(30, 15)) AND l_quantity@0 < Some(2400),15,2                                                                                                           |
|               |             RepartitionExec: partitioning=RoundRobinBatch(20)                                                                                                           |
|               |               ParquetExec: limit=None, partitions=[home/kmitchener/dev/arrow-datafusion/benchmarks/data-parquet/lineitem/part-0.parquet], predicate=l_shipdate_max@0 >= 8766 AND l_shipdate_min@1 < 9131 AND CAST(l_discount_max@2 AS Decimal128(30, 15)) >= CAST(0.049999999999999996 AS Decimal128(30, 15)) AND CAST(l_discount_min@3 AS Decimal128(30, 15)) <= CAST(0.06999999999999999 AS Decimal128(30, 15)) AND l_quantity_min@4 < Some(2400),15,2, projection=[l_quantity, l_extendedprice, l_discount, l_shipdate] |
|               |                                                                                                           |
+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```
**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #3402 .

 # Rationale for this change

Allows the plan to be further optimized and simplified, resulting in better plans.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 1]
```diff
@@ -803,6 +801,26 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
...
+            //
+
+            // a between 3 and 5  -->  a >= 3 AND a <=5
+            // a not between 3 and 5  -->  a < 3 OR a > 5
+            Between {
+                expr,
+                low,
+                high,
+                negated,
+            } => match negated {
```
Comment (Dandandan) [inline datafusion/optimizer/src/simplify_expressions.rs:815]: Could use `if` instead of `match`

File: datafusion/core/tests/sql/predicates.rs [hunk 2]
```diff
@@ -427,11 +427,12 @@ async fn multiple_or_predicates() -> Result<()> {
...
+        "      Filter: #part.p_partkey = #lineitem.l_partkey AND #part.p_brand = Utf8(\"Brand#12\") AND #lineitem.l_quantity >= CAST(Int64(1) AS Float64) AND #lineitem.l_quantity <= CAST(Int64(11) AS Float64) AND #part.p_size <= Int32(5) OR #part.p_brand = Utf8(\"Brand#23\") AND #lineitem.l_quantity >= CAST(Int64(10) AS Float64) AND #lineitem.l_quantity <= CAST(Int64(20) AS Float64) AND #part.p_size <= Int32(10) OR #part.p_brand = Utf8(\"Brand#34\") AND #lineitem.l_quantity >= CAST(Int64(20) AS Float64) AND #lineitem.l_quantity <= CAST(Int64(30) AS Float64) AND #part.p_size <= Int32(15) [l_partkey:Int64, l_quantity:Float64, p_partkey:Int64, p_brand:Utf8, p_size:Int32]",
         "        CrossJoin: [l_partkey:Int64, l_quantity:Float64, p_partkey:Int64, p_brand:Utf8, p_size:Int32]",
         "          TableScan: lineitem projection=[l_partkey, l_quantity] [l_partkey:Int64, l_quantity:Float64]",
-        "          TableScan: part projection=[p_partkey, p_brand, p_size] [p_partkey:Int64, p_brand:Utf8, p_size:Int32]",
+        "          Filter: #part.p_size >= Int32(1) [p_partkey:Int64, p_brand:Utf8, p_size:Int32]",
```
Comment (Dandandan) [inline datafusion/core/tests/sql/predicates.rs:434]: This is a great optimization. Next goal is to get it to see that `part.p_size <= 15` as well :D

**Review by Dandandan — APPROVED** [review]

File: datafusion/core/tests/sql/predicates.rs (hunk 2 above)
Comment (kmitchener) [reply datafusion/core/tests/sql/predicates.rs:434]: Oof, that would be excellent. For another PR if someone doesn't beat me to it :)

Comment (kmitchener) [reply datafusion/core/tests/sql/predicates.rs:434]: Or maybe sooner .. looks like the order of the projections changes between runs, which is causing the test failure here.

**Comment by kmitchener** [comment]
Putting in draft until I can figure out this bug that's triggered by this change. :/

**Comment by Dandandan** [comment]
> Putting in draft until I can figure out this bug that's triggered by this change. :/

Commonly, a hash table or hash set generates these kind of random results. Maybe an optimization rule uses this, where this could be fixed (I e. producing in order of appearance or sorting the results).
You could `explain verbose` to see which rule this is?

**Comment by kmitchener** [comment]
@Dandandan good advice, thank you :)

**Review by Dandandan — APPROVED** [review]
Looks great!

**Review by xudong963 — APPROVED** [review]
LGTM, thanks @kmitchener . I believe the optimization can bring more potential benefits 👍

**Pull request merged**